server: failed to create offline session: not found

Saw this when trying to create a refresh token using dex v2.2.0 through the public clients.

```
time="2017-02-24T01:59:39Z" level=info msg="config issuer: https://eric-different-subnets.staging.core-os.net/identity" 
time="2017-02-24T01:59:39Z" level=info msg="config connector: local passwords enabled" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"auth-code.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"auth-request.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"o-auth2-client.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"signing-key.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"refresh-token.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"password.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=error msg="third party resource already created \"offline-sessions.oidc.coreos.com\"" 
time="2017-02-24T01:59:39Z" level=info msg="config storage: kubernetes" 
time="2017-02-24T01:59:39Z" level=info msg="config static client: tectonic-console" 
time="2017-02-24T01:59:39Z" level=info msg="config static client: tectonic-kubectl" 
time="2017-02-24T01:59:39Z" level=info msg="config skipping approval screen" 
time="2017-02-24T01:59:39Z" level=info msg="listening (http) on 0.0.0.0:5556" 
time="2017-02-24T02:00:38Z" level=error msg="failed to create offline session: not found" 
```

cc @rithujohn191

Conformance tests don't pass for the kubernetes storage on v2.2.0

```
--- FAIL: TestStorage (6.50s)
    --- FAIL: TestStorage/AuthCodeCRUD (0.40s)
    	conformance.go:78: creating an existing auth code expected storage.ErrAlreadyExists, got POST https://eric-different-subnets-k8s.staging.core-os.net:443/apis/oidc.coreos.com/v1/namespaces/default/authcodes Conflict: response from server "{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"thirdpartyresourcedatas.extensions \"smdlx37aok3rnsgkystf6twqsy\" already exists","reason":"AlreadyExists","details":{"name":"smdlx37aok3rnsgkystf6twqsy","group":"extensions","kind":"thirdpartyresourcedatas"},"code":409}"
    --- FAIL: TestStorage/AuthRequestCRUD (0.40s)
    	conformance.go:78: creating an existing auth request expected storage.ErrAlreadyExists, got POST https://eric-different-subnets-k8s.staging.core-os.net:443/apis/oidc.coreos.com/v1/namespaces/default/authrequests Conflict: response from server "{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"thirdpartyresourcedatas.extensions \"tnaznothtuasgpnz72kdssohri\" already exists","reason":"AlreadyExists","details":{"name":"tnaznothtuasgpnz72kdssohri","group":"extensions","kind":"thirdpartyresourcedatas"},"code":409}"
    --- FAIL: TestStorage/ClientCRUD (0.57s)
    	conformance.go:78: creating an existing client expected storage.ErrAlreadyExists, got POST https://eric-different-subnets-k8s.staging.core-os.net:443/apis/oidc.coreos.com/v1/namespaces/default/oauth2clients Conflict: response from server "{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"thirdpartyresourcedatas.extensions \"ojrxa43sgmzgy5lwpfxgu33spbqtmyzuonwhq4dynhf7fhheqqrcgji\" already exists","reason":"AlreadyExists","details":{"name":"ojrxa43sgmzgy5lwpfxgu33spbqtmyzuonwhq4dynhf7fhheqqrcgji","group":"extensions","kind":"thirdpartyresourcedatas"},"code":409}"
    --- FAIL: TestStorage/RefreshTokenCRUD (0.47s)
    	conformance.go:78: creating an existing refresh token expected storage.ErrAlreadyExists, got POST https://eric-different-subnets-k8s.staging.core-os.net:443/apis/oidc.coreos.com/v1/namespaces/default/refreshtokens Conflict: response from server "{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"thirdpartyresourcedatas.extensions \"lhhukvtwa2ffoxhn6aejzvkwem\" already exists","reason":"AlreadyExists","details":{"name":"lhhukvtwa2ffoxhn6aejzvkwem","group":"extensions","kind":"thirdpartyresourcedatas"},"code":409}"
    --- FAIL: TestStorage/PasswordCRUD (0.57s)
    	conformance.go:78: creating an existing password expected storage.ErrAlreadyExists, got POST https://eric-different-subnets-k8s.staging.core-os.net:443/apis/oidc.coreos.com/v1/namespaces/default/passwords Conflict: response from server "{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"thirdpartyresourcedatas.extensions \"njqw4zkamv4gc3lqnrss4y3pnxf7fhheqqrcgji\" already exists","reason":"AlreadyExists","details":{"name":"njqw4zkamv4gc3lqnrss4y3pnxf7fhheqqrcgji","group":"extensions","kind":"thirdpartyresourcedatas"},"code":409}"
    --- PASS: TestStorage/KeysCRUD (0.40s)
    --- FAIL: TestStorage/OfflineSessionCRUD (0.26s)
    	conformance.go:382: create offline session: not found
    --- PASS: TestStorage/GarbageCollection (1.11s)
    --- PASS: TestStorage/TimezoneSupport (0.26s)
    --- PASS: TestStorage/AuthRequestConcurrentUpdate (0.39s)
    --- PASS: TestStorage/ClientConcurrentUpdate (0.45s)
    --- PASS: TestStorage/PasswordConcurrentUpdate (0.39s)
    --- PASS: TestStorage/KeysConcurrentUpdate (0.50s)
```

--- a/storage/kubernetes/client.go
+++ b/storage/kubernetes/client.go
@@ -135,6 +135,9 @@ func checkHTTPErr(r *http.Response, validStatusCodes ...int) error {
 	if r.StatusCode == http.StatusNotFound {
 		return storage.ErrNotFound
 	}
+	if r.Request.Method == "POST" && r.StatusCode == http.StatusConflict {
+		return storage.ErrAlreadyExists
+	}
 
 	var url, method string
 	if r.Request != nil {

--- a/storage/kubernetes/storage.go
+++ b/storage/kubernetes/storage.go
@@ -31,7 +31,7 @@ const (
 	resourceRefreshToken    = "refreshtokens"
 	resourceKeys            = "signingkeies" // Kubernetes attempts to pluralize.
 	resourcePassword        = "passwords"
-	resourceOfflineSessions = "offlinesessions"
+	resourceOfflineSessions = "offlinesessionses" // Again attempts to pluralize.
 )
 
 // Config values for the Kubernetes storage type.

--- a/storage/kubernetes/types.go
+++ b/storage/kubernetes/types.go
@@ -502,9 +502,14 @@ func (cli *client) fromStorageOfflineSessions(o storage.OfflineSessions) Offline
 }
 
 func toStorageOfflineSessions(o OfflineSessions) storage.OfflineSessions {
-	return storage.OfflineSessions{
+	s := storage.OfflineSessions{
 		UserID:  o.UserID,
 		ConnID:  o.ConnID,
 		Refresh: o.Refresh,
 	}
+	if s.Refresh == nil {
+		// Server code assumes this will be non-nil.
+		s.Refresh = make(map[string]*storage.RefreshTokenRef)
+	}
+	return s
 }

